include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1999 @@
+        static_assert(std::is_same<VALUETYPE, deduced_value_type>::value,
+                      "tiny_array_base: type mismatch between VALUETYPE and REPRESENTATION.");
+      public:
+        using base_type              = xbuffer_adaptor<CP, O, A>;
+        using self_type              = xtiny_impl<VALUETYPE, runtime_size, base_type>;
+        using value_type             = VALUETYPE;
+        using const_value_type       = typename std::add_const<value_type>::type;
+        using reference              = value_type &;
+        using const_reference        = const_value_type &;
+        using pointer                = value_type *;
+        using const_pointer          = const_value_type *;
+        using iterator               = typename base_type::iterator;
+        using const_iterator         = typename base_type::const_iterator;
+        using reverse_iterator       = typename base_type::reverse_iterator;
+        using const_reverse_iterator = typename base_type::const_reverse_iterator;
+        using size_type              = std::size_t;
+        using difference_type        = std::ptrdiff_t;
+        using index_type             = std::array<index_t, 1>;
+
+        static const bool owns_memory = false;
+        static const bool has_fixed_size = false;
+        static const index_t static_size = runtime_size;
+
+        using base_type::base_type;
+
+        void assign(size_type n, const value_type& v);
+
+        template <class IT,
+                  class = detail::require_input_iter<IT>>
+        void assign(IT other_begin, IT other_end);
+
+        using base_type::operator[];
+        using base_type::data;
+        using base_type::size;
+
+        constexpr size_type max_size() const;
+        constexpr size_type capacity() const;
+        constexpr index_type shape() const;
+
+        using base_type::begin;
+        using base_type::cbegin;
+        using base_type::rbegin;
+        using base_type::crbegin;
+        using base_type::end;
+        using base_type::cend;
+        using base_type::rend;
+        using base_type::crend;
+
+        using base_type::swap;
+        void swap(xtiny_impl &);
+    };
+
+    template <class V, index_t N, class R>
+    inline void
+    swap(xtiny<V, N, R> & l, xtiny<V, N, R> & r)
+    {
+        l.swap(r);
+    }
+
+    /****************/
+    /* xtiny output */
+    /****************/
+
+    template <class T, index_t N, class R>
+    std::ostream & operator<<(std::ostream & o, xtiny<T, N, R> const & v)
+    {
+        o << "{";
+        if(v.size() > 0)
+            o << promote_type_t<T>(v[0]);
+        for(decltype(v.size()) i=1; i < v.size(); ++i)
+            o << ", " << promote_type_t<T>(v[i]);
+        o << "}";
+        return o;
+    }
+
+    /********************/
+    /* xtiny comparison */
+    /********************/
+
+    template <class V1, index_t N1, class R1, class V2, index_t N2, class R2>
+    inline bool
+    operator==(xtiny<V1, N1, R1> const & l,
```

Comment:
I think the following functions should be modified such that the loops are replaced with `std::equal` and instead of re-implementing almost the same functionality replace with a call to `==` e.g. `!=` -> `!operator==(a, b)` etc.

Reply:
Have you ever benchmarked if `!operator==(a, b)` is as fast as the explicit implementation? I haven't, but the negation is clearly an extra operation that the compiler may or may not optimize away.

The same applies to the call `std::equal(l.begin(), l.end(), r.begin(), r.end())`: Checking if both sequences have equal length requires extra work when `std::equal` is not explicitly specialized for random access iterators. Moreover, my code version will optimize out this check entirely if the array sizes are known at compile time.

Reply:
We have this version in xstorage (uvector).

https://github.com/QuantStack/xtensor/blob/6d6aebbc4f71d436881cf7f82178842a3354c600/include/xtensor/xstorage.hpp#L485-L495

I can do benchmarks tomorrow.

Reply:
I'll wait for your benchmarks before taking action.

Reply:
Here is the generated assembly from GCC 7.2 for both versions, and the std::equal, iterator based version seems to be ~5 instructions smaller: https://godbolt.org/g/cPPhm1

So in effect, the loop is replaced by `memcmp`. I'll do some benchmarks next. The diff between adding a `!` in front is small (it's still using memcmp).

Reply:
I did 2 benchmarks so far: One where all elements are different, and one where all are equal.
The loop based version wins in the case where all elements are different, and doesn't scale great when all elements are the same.

All elements the different (eq1 == std::equal, eq2 == loop based):


```
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
bench_compare_eq1/4             3 ns          3 ns  233995893
bench_compare_eq1/8             3 ns          3 ns  231514648
bench_compare_eq1/64            3 ns          3 ns  235578626
bench_compare_eq1/512           3 ns          3 ns  234599273
bench_compare_eq1/1000          3 ns          3 ns  234425023
bench_compare_eq2/4             1 ns          1 ns  629091844
bench_compare_eq2/8             1 ns          1 ns  625105046
bench_compare_eq2/64            1 ns          1 ns  629999371
bench_compare_eq2/512           1 ns          1 ns  627157658
bench_compare_eq2/1000          1 ns          1 ns  616530794
```

All elements the same:

```
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
bench_compare_eq1/4             3 ns          3 ns  234980526
bench_compare_eq1/8             3 ns          3 ns  232474783
bench_compare_eq1/64           10 ns         10 ns   73541694
bench_compare_eq1/512          48 ns         48 ns   13767823
bench_compare_eq1/1000         89 ns         89 ns    7686126
bench_compare_eq2/4             3 ns          3 ns  205210607
bench_compare_eq2/8             6 ns          6 ns  121201197
bench_compare_eq2/64           36 ns         36 ns   19272220
bench_compare_eq2/512         291 ns        291 ns    2416103
bench_compare_eq2/1000        555 ns        554 ns    1183632
```

Reply:
I've uploaded the benchmark here: https://gist.github.com/wolfv/a64526ceb94dd37b4a5f2f5dbcf76779 maybe you have time to check the results on MSVC?

I've modified the benchmark so that it modifies a random index each loop in an otherwise equal array to create some variance in the distance where the arrays are not equal. This incurs a heavy cost though ...

These are the latest results:

```
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
bench_compare_eq1/4            77 ns         77 ns    8853946
bench_compare_eq1/8            78 ns         78 ns    8857423
bench_compare_eq1/64           91 ns         91 ns    7479546
bench_compare_eq1/512         183 ns        183 ns    3750081
bench_compare_eq1/1000        273 ns        273 ns    2567153
bench_compare_eq2/4            77 ns         77 ns    8866009
bench_compare_eq2/8            79 ns         79 ns    8777499
bench_compare_eq2/64          112 ns        112 ns    6227843
bench_compare_eq2/512         350 ns        350 ns    2027685
bench_compare_eq2/1000        588 ns        588 ns    1163340
```

Reply:
Great! I'll run your benchmarks later today.

Reply:
Just fyi I found this implementation

```
auto ita = a.begin();
auto itb = b.begin();
auto resit = result.begin();
auto n = std::distance(ita, a.end());
for (std::size_t i = 0; i < n; ++i)
{
    *resit = (*ita + *itb);
    ++resit;
    ++ita;
    ++itb;
}
```

on std::arrays to be faster than the integer-access based (gcc 7.2). The speed difference was 3ns vs 1ns.

Reply:
Here are the benchmark results on Windows:
```
Run on (8 X 1993 MHz CPU s), Visual Studio 2015
SIMD SIZE: 2
----------------------------------------------------------------
Benchmark                         Time           CPU Iterations
----------------------------------------------------------------
bench_compare_eq1/4             507 ns        479 ns    1725827
bench_compare_eq1/8             390 ns        389 ns    1725827
bench_compare_eq1/64            415 ns        413 ns    1547293
bench_compare_eq1/512           829 ns        834 ns     897430
bench_compare_eq1/1000         1506 ns       1502 ns     498572
bench_compare_eq2/4             347 ns        348 ns    1794860
bench_compare_eq2/8             399 ns        392 ns    1869646
bench_compare_eq2/64            544 ns        534 ns    1547293
bench_compare_eq2/512          1204 ns       1189 ns     498572
bench_compare_eq2/1000         1844 ns       1877 ns     373929
bench_compare_fill/4              7 ns          7 ns  112178768
bench_compare_fill/8             10 ns         10 ns   64102153
bench_compare_fill/64            58 ns         57 ns   11217877
bench_compare_fill/512          307 ns        307 ns    2136738
bench_compare_fill/1000         579 ns        577 ns    1000000
bench_compare_fill2/4            14 ns         15 ns   44871507
bench_compare_fill2/8            15 ns         15 ns   44871507
bench_compare_fill2/64           41 ns         41 ns   18696461
bench_compare_fill2/512         273 ns        273 ns    2804469
bench_compare_fill2/1000        508 ns        515 ns    1000000
----------------------------------------------------------------
SIMD SIZE: 4
----------------------------------------------------------------
Benchmark                         Time           CPU Iterations
----------------------------------------------------------------
bench_compare_eq1/4             416 ns        365 ns    1794860
bench_compare_eq1/8             371 ns        368 ns    1950935
bench_compare_eq1/64            407 ns        409 ns    1869646
bench_compare_eq1/512           791 ns        782 ns     897430
bench_compare_eq1/1000         1382 ns       1391 ns     560894
bench_compare_eq2/4             357 ns        352 ns    2039614
bench_compare_eq2/8             366 ns        368 ns    1950935
bench_compare_eq2/64            449 ns        441 ns    1661908
bench_compare_eq2/512          1063 ns       1046 ns     641022
bench_compare_eq2/1000         1638 ns       1606 ns     407923
bench_compare_fill/4              7 ns          7 ns  112178768
bench_compare_fill/8             10 ns         10 ns   74785845
bench_compare_fill/64            56 ns         57 ns   11217877
bench_compare_fill/512          268 ns        269 ns    2492862
bench_compare_fill/1000         511 ns        515 ns    1000000
bench_compare_fill2/4            15 ns         15 ns   44871507
bench_compare_fill2/8            14 ns         14 ns   49857230
bench_compare_fill2/64           37 ns         38 ns   18696461
bench_compare_fill2/512         242 ns        245 ns    2991434
bench_compare_fill2/1000        533 ns        528 ns    1447468
```

Reply:
changed